compiler/arser/tests/arser.test.cpp

```diff
@@ -0,0 +1,179 @@
+  char **argv(void) { return _argv.data(); }
+
+private:
+  std::vector<char *> _argv;
+  std::vector<std::string> _arg;
+};
+
+TEST(BasicTest, option)
+{
+  /* arrange */
+  Arser arser;
+
+  arser.add_argument("--verbose")
+      .nargs(0)
+      .help("It provides additional details as to what the executable is doing");
+
+  Prompt prompt("./executable --verbose");
+  /* act */
+  arser.parse(prompt.argc(), prompt.argv());
+  /* assert */
+  EXPECT_TRUE(arser["--verbose"]);
+  EXPECT_TRUE(arser.get<bool>("--verbose"));
+}
+
+TEST(BasicTest, OptionalArgument)
+{
+  /* arrange */
+  Arser arser;
+
+  arser.add_argument("--volume")
+      .nargs(1)
+      .type(arser::DataType::INT32)
+      .help("Set a volume as you provided.");
+
+  Prompt prompt("./radio --volume 5");
+  /* act */
+  arser.parse(prompt.argc(), prompt.argv());
+  /* assert */
+  EXPECT_TRUE(arser["--volume"]);
+  EXPECT_EQ(5, arser.get<int>("--volume"));
+
+  EXPECT_FALSE(arser["--price"]);
+  EXPECT_THROW(arser.get<bool>("--volume"), std::runtime_error);
+}
+
+TEST(BasicTest, NonRequiredOptionalArgument)
+{
+  /* arrange */
+  Arser arser;
+
+  arser.add_argument("--weight")
+      .nargs(1)
+      .type(arser::DataType::INT32)
+      .help("Set a volume as you provided.");
+
+  Prompt prompt("./radio"); // empty argument
+  /* act */
+  arser.parse(prompt.argc(), prompt.argv());
+  /* assert */
+  EXPECT_FALSE(arser["--volume"]);
+  EXPECT_THROW(arser.get<int>("--weight"), std::runtime_error);
+}
+
+TEST(BasicTest, RequiredOptionalArgument)
+{
+  /* arrange */
+  Arser arser;
+
+  arser.add_argument("--volume")
+      .nargs(1)
+      .type(arser::DataType::INT32)
+      .required()
+      .help("Set a volume as you provided.");
+
+  Prompt prompt("./radio");
+  /* act */ /* assert */
+  EXPECT_THROW(arser.parse(prompt.argc(), prompt.argv()), std::runtime_error);
+}
+
+TEST(BasicTest, OptionalMultipleArgument)
+{
+  /* arrange */
+  Arser arser;
+
+  arser.add_argument("--add").nargs(2).type(arser::DataType::INT32_VEC).help("Add two numbers.");
+
+  Prompt prompt("./calculator --add 3 5");
+  /* act */
+  arser.parse(prompt.argc(), prompt.argv());
+  /* assert */
+  EXPECT_TRUE(arser["--add"]);
+  std::vector<int> values = arser.get<std::vector<int>>("--add");
+  EXPECT_EQ(3, values.at(0));
+  EXPECT_EQ(5, values.at(1));
+
+  EXPECT_THROW(arser.get<std::vector<float>>("--add"), std::runtime_error);
+}
+
+TEST(BasicTest, MultipleOptionalArgument)
+{
+  /* arrange */
+  Arser arser;
+
+  arser.add_argument("--input_path")
+      .nargs(1)
+      .type(arser::DataType::STR)
+      .help("input path of this program.")
+      .required();
+  arser.add_argument("--output_path")
+      .nargs(1)
+      .type(arser::DataType::STR)
+      .help("output path of this program.")
+      .required(true);
+  arser.add_argument("--training_data")
+      .nargs(5)
+      .type(arser::DataType::INT32_VEC)
+      .help("give traning data to this program.")
+      .required();
```

Comment:
Looks like a C++ version of Python argparse. Am I right?

Reply:
Rigth:) That looks simple.